File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int n = fruits.size();
        unordered_map<int, int> baskets;

        int l=0, r=-1, maxNum=0;
        while(++r<n){
            baskets[fruits[r]]++;
            while(baskets.size()>2){
                baskets[fruits[l]]--;
                if(baskets[fruits[l]]==0) baskets.erase(fruits[l]);
                l++;
            }
            if(r-l+1>maxNum)  maxNum=r-l+1;
        }
        return maxNum;
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        n = len(fruits)
        count = defaultdict(int)
        left = 0
        max_fruits = 0
        for right in range(n):
            count[fruits[right]]+=1
            while left < n and len(count.keys()) > 2:
                count[fruits[left]] -= 1
                if count[fruits[left]] == 0:
                    count.pop(fruits[left])
                left += 1
            max_fruits = max(max_fruits, right-left+1)
        return max_fruits